app/code/Magento/Reports/Controller/Adminhtml/Report/Shopcart.php

```diff
@@ -54,19 +59,38 @@ public function _initAction()
      */
     protected function _isAllowed()
     {
-        switch ($this->getRequest()->getActionName()) {
+        $actionName = strtolower($this->getRequest()->getActionName());
```

Comment:
I do not like this approach because we are still supporting violation of Dependency Inversion principle.
In current place Shopcart depends on it's children classes.
In any time we will add new child in hierarchy, we will be forced to change our abstract class.

I think it will be better to use constant \Magento\Backend\App\AbstractAction::ADMIN_RESOURCE and override it in each child.

E.g. 
\Magento\Reports\Controller\Adminhtml\Report\Shopcart::ADMIN_RESOURCE = 'Magento_Reports::shopcart';
\Magento\Reports\Controller\Adminhtml\Report\Shopcart\Product::ADMIN_RESOURCE = 'Magento_Reports::product';
\Magento\Reports\Controller\Adminhtml\Report\Shopcart\Customer::ADMIN_RESOURCE = null;
Etc.

Reply:
Hello @swnsma : Thank you for your feedback. I know it but some Backward compatible development rule prevent me to change it. It will make the "Health Test", or "Static Test", etc...
FAIL.
So it is very difficult to change the source code.

Reply:
@swnsma : Magento\Reports\Controller\Adminhtml\Report\Shopcart has the bad structure because it is abstract but it process the child class name.  However, I can not change the source code because it is marked as @api .
Do you know how to change the source code and respect the rule? Could I ask someone of Magento to change it.
Thank you very much.

Reply:
@edenduong
Could specify which rules are violated in your case?
I [docs](https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/#php-1) I can see that changing constant value is allowed action in scope of backward compatibility.

Reply:
@swnsma: I don't remember. I will change the code and wait the report again :)  Thank you.

Reply:
But I seems we must ADD the Constant, not only CHANGE it. So it is the problem.

Reply:
It is difficult to say that we ADD constant.
We already have it in parent class and we only do - change the constant for the child.
I will discuss this question with other maintainers.

Reply:
![image](https://user-images.githubusercontent.com/8234209/62132082-9277a100-b306-11e9-94e5-6b7d43b354f2.png)

Reply:
But I think I can pass it in the "isAllow" function and don't need add it. Hope it can pass.